### PR TITLE
Transaction controller refractor pt3

### PR DESCRIPTION
**Pull request by frankiebee**
- [x] break out state managment and store into the tx-state-manager
- [x] all tx files should have at least %90 test coverage 
- [x] transaction controller should be easy to read and easy to test for

### CODE REVIEW

File: app/scripts/controllers/transactions.js [hunk 1]
```diff
@@ -2,51 +2,62 @@ const EventEmitter = require('events')
...
     this.networkStore = opts.networkStore || new ObservableStore({})
     this.preferencesStore = opts.preferencesStore || new ObservableStore({})
-    this.txHistoryLimit = opts.txHistoryLimit
     this.provider = opts.provider
     this.blockTracker = opts.blockTracker
     this.signEthTx = opts.signTransaction
     this.ethStore = opts.ethStore
 
+    this.memStore = new ObservableStore({})
+    this.query = new EthQuery(this.provider)
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: do we want to use `ethjs-rpc` instead of `ethjs-query` (no formatting, pending promisification of `ethjs-rpc`)

File: app/scripts/controllers/transactions.js [hunk 2]
```diff
@@ -2,51 +2,62 @@ const EventEmitter = require('events')
...
+    this.txGasUtil = new TxGasUtil(this.provider)
+
+    this.txStateManager = new TransactionStateManger({
+      initState: extend({
+        transactions: [],
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: default state should go inside `TransactionStateManger`

File: app/scripts/controllers/transactions.js [hunk 3]
```diff
@@ -56,19 +67,22 @@ module.exports = class TransactionController extends EventEmitter {
...
-        const network = this.getNetwork()
-        return this.getFilteredTxList({
-          status: 'submitted',
-          metamaskNetworkId: network,
-        })
+      publishTransaction: this.query.sendRawTransaction,
+      getPendingTransactions: this.txStateManager.getPendingTransactions.bind(this.txStateManager),
+      giveUpOnTransaction: (txId) => {
+        const msg = `Gave up submitting after 3500 blocks un-mined.`
+        this.setTxStatusFailed(txId, msg)
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: is string correct or do we need to give it an err obj

File: app/scripts/controllers/transactions.js [hunk 4]
```diff
@@ -56,19 +67,22 @@ module.exports = class TransactionController extends EventEmitter {
...
       },
     })
 
-    this.pendingTxTracker.on('txWarning', this.updateTx.bind(this))
-    this.pendingTxTracker.on('txFailed', this.setTxStatusFailed.bind(this))
-    this.pendingTxTracker.on('txConfirmed', this.setTxStatusConfirmed.bind(this))
+    this.txStateManager.subscribe(() => {
+      this.emit('update')
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: dont need to emit update if our memStore is updated -- consumers should observe memStore

File: app/scripts/controllers/transactions.js [hunk 5]
```diff
@@ -80,7 +94,7 @@ module.exports = class TransactionController extends EventEmitter {
     this.blockTracker.on('sync', this.pendingTxTracker.queryPendingTxs.bind(this.pendingTxTracker))
     // memstore is computed from a few different stores
     this._updateMemstore()
-    this.store.subscribe(() => this._updateMemstore())
+    this.txStateManager.subscribe(() => this._updateMemstore())
     this.networkStore.subscribe(() => this._updateMemstore())
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: these should prolly be composed into the parent memStore

File: app/scripts/lib/tx-state-manager.js [hunk 6]
```diff
@@ -0,0 +1,230 @@
+const extend = require('xtend')
+const ObservableStore = require('obs-store')
+const txStateHistoryHelper = require('./tx-state-history-helper')
+
+module.exports = class TransactionStateManger extends ObservableStore {
```
Comment (kumavis) [inline app/scripts/lib/tx-state-manager.js]: i dont think if we should extend obs-store or just have one
the pattern so far is to have the controller and its memStore

File: app/scripts/lib/tx-state-manager.js [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+      stack: err.stack,
+    }
+    this.updateTx(txMeta)
+    this._setTxStatus(txId, 'failed')
+  }
+
+/* _____________________________________
+|                                      |
+|           PRIVATE METHODS            |
+|______________________________________*/
```
Comment (kumavis) [inline app/scripts/lib/tx-state-manager.js]: our code style for comment sections in other places has been
```
//
/// title
//
```

**Review by kumavis — CHANGES_REQUESTED** [review]
nice to see it broken up
theres some architectural inconsistencies (controller pattern vs obs-store subclass)
generally good, excited for the improved test coverage

File: app/scripts/controllers/transactions.js (hunk 1 above)
Comment (kumavis) [reply app/scripts/controllers/transactions.js]: update: `ethjs-rpc@0.1.7` is promised based :+1:

Comment (frankiebee) [reply app/scripts/controllers/transactions.js]: could be into that.. will take a look

File: app/scripts/controllers/transactions.js (hunk 3 above)
Comment (frankiebee) [reply app/scripts/controllers/transactions.js]: this dosent look like something i wrote must have come from merging master the other day

Comment (frankiebee) [reply app/scripts/controllers/transactions.js]: but ill fix it

File: app/scripts/lib/tx-state-manager.js (hunk 7 above)
Comment (frankiebee) [reply app/scripts/lib/tx-state-manager.js]: so this is from a while ago i made this big and bold because their were some non private looking functions in the private section so this was a little louder but if you are so against it ill change it

File: app/scripts/controllers/transactions.js (hunk 2 above)
Comment (frankiebee) [reply app/scripts/controllers/transactions.js]: ✔️

File: app/scripts/controllers/transactions.js (hunk 3 above)
Comment (frankiebee) [reply app/scripts/controllers/transactions.js]: ✔️

File: app/scripts/controllers/transactions.js (hunk 4 above)
Comment (frankiebee) [reply app/scripts/controllers/transactions.js]: ✔️

File: app/scripts/lib/tx-state-manager.js (hunk 6 above)
Comment (frankiebee) [reply app/scripts/lib/tx-state-manager.js]: ✔️

File: app/scripts/lib/tx-state-manager.js (hunk 7 above)
Comment (frankiebee) [reply app/scripts/lib/tx-state-manager.js]: ✔️

**Comment by frankiebee** [comment]
Awating @tmashuang push for tests

File: app/scripts/lib/tx-state-manager.js [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    const txCount = this.getTxCount()
+    const txHistoryLimit = this.txHistoryLimit
+
+    // checks if the length of the tx history is
+    // longer then desired persistence limit
+    // and then if it is removes only confirmed
+    // or rejected tx's.
+    // not tx's that are pending or unapproved
+    if (txCount > txHistoryLimit - 1) {
+      const index = transactions.findIndex((metaTx) => ((metaTx.status === 'confirmed' || metaTx.status === 'rejected')))
```
Comment (kumavis) [inline app/scripts/lib/tx-state-manager.js]: extra parens ( doesnt need double )

File: app/scripts/metamask-controller.js [hunk 9]
```diff
@@ -133,7 +133,7 @@ module.exports = class MetamaskController extends EventEmitter {
     this.publicConfigStore = this.initPublicConfigStore()
 
     // manual disk state subscriptions
-    this.txController.store.subscribe((state) => {
+    this.txController.txStateManager.store.subscribe((state) => {
```
Comment (kumavis) [inline app/scripts/metamask-controller.js]: lets actually subscribe to the `txController` store. if its doesnt have one, it should set the `txStateManager`'s store as its own

File: test/unit/components/pending-tx-test.js [hunk 10]
```diff
@@ -24,7 +24,8 @@ describe('PendingTx', function () {
       'to': '0xc5b8dbac4c1d3f152cdeb400e2313f309c410acb',
       'value': '0xde0b6b3a7640000',
       gasPrice,
-      'gas': '0x7b0c'},
+      'gas': '0x7b0c',
+    },
```
Comment (kumavis) [inline test/unit/components/pending-tx-test.js]: 😸

File: test/unit/tx-utils-test.js [hunk 11]
```diff
@@ -28,7 +30,7 @@ describe('txUtils', function () {
         nonce: '0x3',
         chainId: 42,
       }
-      const ethTx = txUtils.buildEthTxFromParams(txParams)
+      const ethTx = new Transaction(txParams)
```
Comment (kumavis) [inline test/unit/tx-utils-test.js]: yay!

File: test/unit/tx-controller-test.js [hunk 12]
```diff
@@ -34,27 +35,83 @@ describe('Transaction Controller', function () {
...
+        { id: 1, status: 'submitted', metamaskNetworkId: currentNetworkId, txParams: {} },
+        { id: 2, status: 'submitted', metamaskNetworkId: currentNetworkId, txParams: {} },
+        { id: 3, status: 'submitted', metamaskNetworkId: currentNetworkId, txParams: {} },
+      ])
+      const pendingTxCount = txController.getPendingTxCount()
+      assert.equal(pendingTxCount, 3, 'should be 3')
+    })
+  })
+
+  describe('#getConfirmedTransactions', function () {
```
Comment (frankiebee) [inline test/unit/tx-controller-test.js]: @tmashuang this should go in the appropriate file

File: test/unit/tx-controller-test.js [hunk 13]
```diff
@@ -286,4 +315,143 @@ describe('Transaction Controller', function () {
...
+      assert.equal(txController.getChainId(), 0)
+    })
+  })
+
+  describe('#publishTransaction', async function () {
+    beforeEach(function () {
+      const txMeta = [
+        { id: 0, status: 'unapproved', txParams: { from: '0x1678a085c290ebd122dc42cba69373b5953b831d', nonce: '0x1', value: '0xfffff' }, rawTx: 'f8498080808080801ca00255b75b550cf112e18fd699f27d043d85348c29d7e8fd234799db890f7c272da0238121aed40c3141e63aae5335aaa3c9711d4907f28071f81f8d055b9f8435e0', metamaskNetworkId: currentNetworkId },
+      ]
+      txController.txStateManager.addTx(txMeta)
```
Comment (tmashuang) [inline test/unit/tx-controller-test.js]: replace addTx with _saveTxList

File: test/unit/tx-controller-test.js [hunk 14]
```diff
@@ -286,4 +315,143 @@ describe('Transaction Controller', function () {
...
+    })
+
+    it('should update rawTx of a transaction', async function () {
+      txController.publishTransaction(0, 'f84c01808080830fffff801ca0e23554ce6f82186402dcdcfff377793fdfa8a872a5670c0ffc002c9cd2f6827aa02a83dfce20aef4064a55320bda47394043af3cbfa63c4e029c54ba6281dcc70e')
```
Comment (tmashuang) [inline test/unit/tx-controller-test.js]: stub publishTransaction from provider

File: test/unit/tx-controller-test.js [hunk 15]
```diff
@@ -286,4 +315,143 @@ describe('Transaction Controller', function () {
...
+      const txMetas = [
+        { id: 0, status: 'unapproved', txParams: { }, metamaskNetworkId: currentNetworkId },
+        { id: 1, status: 'rejected', txParams: { }, metamaskNetworkId: currentNetworkId },
+        { id: 2, status: 'approved', txParams: { }, metamaskNetworkId: currentNetworkId },
+        { id: 3, status: 'signed', txParams: { }, metamaskNetworkId: currentNetworkId },
+        { id: 4, status: 'submitted', txParams: { }, metamaskNetworkId: currentNetworkId },
+        { id: 5, status: 'confirmed', txParams: { }, metamaskNetworkId: currentNetworkId },
+        { id: 6, status: 'failed', txParams: { }, metamaskNetworkId: currentNetworkId },
+      ]
+      txMetas.forEach((txMeta) => txController.txStateManager.addTx(txMeta))
```
Comment (tmashuang) [inline test/unit/tx-controller-test.js]: replace addTx with _saveTxList

File: app/scripts/controllers/transactions.js [hunk 16]
```diff
@@ -193,7 +189,7 @@ module.exports = class TransactionController extends EventEmitter {
       // wait for a nonce
       nonceLock = await this.nonceTracker.getNonceLock(fromAddress)
       // add nonce to txParams
-      txMeta.txParams.nonce = nonceLock.nextNonce
+      txMeta.txParams.nonce = ethUtil.addHexPrefix(nonceLock.nextNonce.toString(16))
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: is there a `ethUtil.intToHex`? if not, there should be

File: app/scripts/controllers/transactions.js [hunk 17]
```diff
@@ -60,17 +59,14 @@ module.exports = class TransactionController extends EventEmitter {
...
       provider: this.provider,
       nonceTracker: this.nonceTracker,
+      retryLimit: 3500, // Retry 3500 blocks, or about 1 day.
       getBalance: (address) => {
         const account = this.ethStore.getState().accounts[address]
         if (!account) return
         return account.balance
       },
-      publishTransaction: this.query.sendRawTransaction,
+      publishTransaction: (rawTx) => this.query.sendRawTransaction(rawTx),
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: why do we do this?
seems like `PendingTransactionTracker` should use its own query around provider to submit the tx

File: app/scripts/controllers/transactions.js [hunk 18]
```diff
@@ -70,15 +68,12 @@ module.exports = class TransactionController extends EventEmitter {
       publishTransaction: this.query.sendRawTransaction,
       getPendingTransactions: this.txStateManager.getPendingTransactions.bind(this.txStateManager),
       giveUpOnTransaction: (txId) => {
-        const msg = `Gave up submitting after 3500 blocks un-mined.`
-        this.setTxStatusFailed(txId, msg)
+        const err = new Error(`Gave up submitting after 3500 blocks un-mined.`)
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: can't hard code this number, need to read `provider` from config

File: app/scripts/controllers/transactions.js [hunk 19]
```diff
@@ -250,10 +245,9 @@ module.exports = class TransactionController extends EventEmitter {
     this.txStateManager.updateTx(txMeta)
   }
 
-/* _____________________________________
-|                                      |
-|           PRIVATE METHODS            |
-|______________________________________*/
+//
+//           PRIVATE METHODS
+//
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: ty

File: test/unit/tx-controller-test.js [hunk 20]
```diff
@@ -286,4 +315,143 @@ describe('Transaction Controller', function () {
...
+  describe('#publishTransaction', function () {
+    let replaceRawTx, rawTx, hash, txMeta
+    beforeEach(function () {
+      rawTx = 'f86c808504a817c800827b0d940c62bb85faa3311a998d3aba8098c1235c564966880de0b6b3a7640000802aa08ff665feb887a25d4099e40e11f0fef93ee9608f404bd3f853dd9e84ed3317a6a02ec9d3d1d6e176d4d2593dd760e74ccac753e6a0ea0d00cc9789d0d7ff1f471d'
+      replaceRawTx = 'f84c01808080830fffff801ca0e23554ce6f82186402dcdcfff377793fdfa8a872a5670c0ffc002c9cd2f6827aa02a83dfce20aef4064a55320bda47394043af3cbfa63c4e029c54ba6281dcc70e'
+      txMeta = {
+        id: 1,
+        status: 'approved',
+        txParams: {},
+        rawTx,
```
Comment (tmashuang) [inline test/unit/tx-controller-test.js]: remove rawTx

File: app/scripts/controllers/transactions.js (hunk 5 above)
Comment (kumavis) [reply app/scripts/controllers/transactions.js]: lets see if we can get these to compose better

File: app/scripts/lib/tx-state-manager.js [hunk 21]
```diff
@@ -180,9 +180,12 @@ module.exports = class TransactionStateManger extends ObservableStore {
     this._setTxStatus(txId, 'confirmed')
   }
 
-  setTxStatusFailed (txId, reason) {
+  setTxStatusFailed (txId, err) {
```
Comment (kumavis) [inline app/scripts/lib/tx-state-manager.js]: 👌 👍

**Review by kumavis — CHANGES_REQUESTED** [review]
there were diffs and a file move so it was hard to follow -- please avoid that in the future
fun fact: thats how the parity multisig bug was introduced

went through some of the commits one at a time, so i may have a distorted sense of how it all fits together

main take aways:
- small style change requests
- lets do better obs-store composing

File: app/scripts/metamask-controller.js (hunk 9 above)
Comment (frankiebee) [reply app/scripts/metamask-controller.js]: 👍

**Comment by kumavis** [comment]
we are modifying txMeta state in the pending-tx-tracker without saving the state. in general we shouldnt be modifying state there if we can avoid it

**Comment by kumavis** [comment]
theres still a DO NOT MERGE label on this

File: app/scripts/controllers/transactions.js (hunk 16 above)
Comment (wanderer) [reply app/scripts/controllers/transactions.js]: i don't think ethereumjs-util should replicate javascript. just do '0x' + int.toString(16) :smile_cat:

Comment (kumavis) [reply app/scripts/controllers/transactions.js]: yeah but we need type validation
```
> `0x${(NaN).toString(16)}`
'0xNaN'
```

**Review by kumavis — APPROVED** [review]

**Comment by kumavis** [comment]
@frankiebee good work here, nice cleanup
